### PR TITLE
Issue178 code coverage

### PR DESCRIPTION
**Pull request by genadyp**
Committed code includes:
-  Code coverage support

```
Creates coverage report based on rspecs.
Usage:
   rake coverage
Current implementation: 
   SimpleCov package.
SimpleCov suggestions:
   * see spec/spec_helper.rb file
   * SimpleCov related code must be issued before any of your application code is required
Output:
   <project root>/coverage directory, containing report files, will be created.
To view report:
   open in a browser <project root>/coverage/index.html file. 
```
- move all minitest tests to RSpec format
- new_spec rake task, that creates new template spec

```
Creates new template rspec file associated with provided source file.
Parameter:
  src=<path of associated source file>
  If no parameter was provided, then user will be asked to enter a source file.
Usage:  
     rake new_spec src=<path of associated source file>
     rake new_spec
     Examples:
       rake new_spec src=lib/email/email.rb
       rake new_spec
Output:
  New template rspec file will be created.
  If absent then a suite directory will be created (see examples below).
  Example:
    * rake new_spec src=lib/email/email.rb
      Suite directory will be created: spec/email/          
      RSpec file will be created: spec/email/email_spec.rb
    * rake new_spec src=lib/file_monitoring/file_monitoring.rb   
      We already have spec/file_monitoring/ directory, so no need to create new one.
      RSpec file will be created: spec/file_monitoring/file_monitoring_spec.rb
```
#### Rake from master:

```
[genadyp@lnx-genadyp bbfs]$ rake
/home/genadyp/.rvm/rubies/ruby-1.9.3-p327/bin/ruby -I"lib:test:spec" -I"/home/genadyp/.rvm/gems/ruby-1.9.3-p327/gems/rake-10.1.0/lib" "/home/genadyp/.rvm/gems/ruby-1.9.3-p327/gems/rake-10.1.0/lib/rake/rake_test_loader.rb" "./test/content_data/content_data_test.rb" "./test/file_monitoring/monitor_path_test.rb" "./test/file_utils/fileutil_mksymlink_test.rb" "./test/file_utils/time_modification_test.rb" "./test/params/params_test.rb" "./test/run_in_background/run_in_background_test.rb" 
Run options: 

# Running tests:

.................daemon/service good_daemon_test started
daemon/service good_daemonize_test started
good_daemon_test: trying to stop process with pid 22399...
good_daemon_test: process with pid 22399 successfully stopped.
daemon/service good_daemon_test deleted
good_daemonize_test: trying to stop process with pid 22411...
good_daemonize_test: process with pid 22411 successfully stopped.
daemon/service good_daemonize_test deleted
.

Finished tests in 3.224787s, 5.5818 tests/s, 29.7694 assertions/s.

18 tests, 96 assertions, 0 failures, 0 errors, 0 skips
/home/genadyp/.rvm/rubies/ruby-1.9.3-p327/bin/ruby -S rspec ./spec/content_data/validations_spec.rb ./spec/content_server/content_server_spec.rb ./spec/content_server/file_streamer_spec.rb ./spec/file_copy/copy_spec.rb ./spec/file_indexing/index_agent_spec.rb ./spec/file_monitoring/monitor_path_spec.rb ./spec/networking/tcp_spec.rb ./spec/validations/index_validations_spec.rb ./test/file_generator/file_generator_spec.rb ./test/params/params_spec.rb
.......................DEPRECATION: mock is deprecated. Use double instead. Called from /home/genadyp/Dev/BBFS/bbfs/test/file_generator/file_generator_spec.rb:22:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /home/genadyp/Dev/BBFS/bbfs/test/file_generator/file_generator_spec.rb:35:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /home/genadyp/Dev/BBFS/bbfs/test/file_generator/file_generator_spec.rb:49:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /home/genadyp/Dev/BBFS/bbfs/test/file_generator/file_generator_spec.rb:62:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /home/genadyp/Dev/BBFS/bbfs/test/file_generator/file_generator_spec.rb:76:in `block (2 levels) in <module:Spec>'.
................

Finished in 0.56183 seconds
43 examples, 0 failures
[genadyp@lnx-genadyp bbfs]$ 
```
#### Rake from branch:

```
[genadyp@f1 bbfs]$ rake
/usr/local/rvm/rubies/ruby-1.9.3-p484/bin/ruby -S rspec ./spec/content_data/content_data_spec.rb ./spec/content_data/validations_spec.rb ./spec/content_server/content_server_spec.rb ./spec/content_server/file_streamer_spec.rb ./spec/file_copy/copy_spec.rb ./spec/file_generator/file_generator_spec.rb ./spec/file_indexing/index_agent_spec.rb ./spec/file_monitoring/monitor_path_spec.rb ./spec/file_utils/fileutil_mksymlink_spec.rb ./spec/file_utils/time_modification_spec.rb ./spec/networking/tcp_spec.rb ./spec/params/params_spec.rb ./spec/run_in_background/run_in_background_spec.rb ./spec/validations/index_validations_spec.rb
..........................DEPRECATION: mock is deprecated. Use double instead. Called from /tmp/genadyp/bbfs/spec/file_generator/file_generator_spec.rb:27:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /tmp/genadyp/bbfs/spec/file_generator/file_generator_spec.rb:40:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /tmp/genadyp/bbfs/spec/file_generator/file_generator_spec.rb:54:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /tmp/genadyp/bbfs/spec/file_generator/file_generator_spec.rb:67:in `block (2 levels) in <module:Spec>'.
.DEPRECATION: mock is deprecated. Use double instead. Called from /tmp/genadyp/bbfs/spec/file_generator/file_generator_spec.rb:81:in `block (2 levels) in <module:Spec>'.
..........................daemon/service good_daemon_test started
daemon/service good_daemonize_test started
good_daemon_test: trying to stop process with pid 21086...
good_daemon_test: process with pid 21086 successfully stopped.
daemon/service good_daemon_test deleted
good_daemonize_test: trying to stop process with pid 21098...
good_daemonize_test: process with pid 21098 successfully stopped.
daemon/service good_daemonize_test deleted
.....

Finished in 4.22 seconds
61 examples, 0 failures
```

Number of minitests + RSpec in master = number of RSpec in the branch.
#### Review directions:

Actually only Rakefile and, optionally, .simplecov should be reviewed.

@yarondbb 
@kolmanv 
Do we want this branch to be merge before the release?

### CODE REVIEW

**Comment by genadyp** [comment]
@kolmanv 
@yarondbb 
@AlexeyNemytov 
Review, please

**Comment by genadyp** [comment]
Pull request connected to #178

File: .simplecov [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  SimpleCov.merge_timeout 3600
+  SimpleCov.start do
+    # filter out files from the spec directory,
+    # cause we are interested in coverage report for source files
+    add_filter "/spec/"
+
+    # filter out source files that do not have method definition
+    # this makes report more clear
+    add_filter do |source_file|
+      method_line_index = source_file.lines.index { |line| line.src =~ /^\s*def / }
```
Comment (bbfsdev) [inline .simplecov:12]: Can you give examples for such (files without def) files?

File: Rakefile [hunk 2]
```diff
@@ -25,11 +26,120 @@ task :uninstall do
...
+#   New template rspec file will be created.
+#   If absent then a suite directory will be created (see examples below).
+#   Example:
+#     * rake new_spec src=lib/email/email.rb
+#       Suite directory will be created: spec/email/    
+#       RSpec file will be created: spec/email/email_spec.rb
+#     * rake new_spec src=lib/file_monitoring/file_monitoring.rb  
+#       We already have spec/file_monitoring/ directory, so no need to create new one.
+#       RSpec file will be created: spec/file_monitoring/file_monitoring_spec.rb
+desc 'Add new spec template'
```
Comment (bbfsdev) [inline Rakefile:66]: You really should update the whole team how to use this.
This should be communicated to everyone.

File: Rakefile [hunk 3]
```diff
@@ -25,11 +26,120 @@ task :uninstall do
...
+  
+  require 'pathname'
+  src_abs_pathname = Pathname.new(File.absolute_path(src_path))
+  # suggestion that Rakefile under the root dir of the project
+  project_abs_pathname = Pathname.new(File.absolute_path(File.dirname(__FILE__)))
+  src_rel_pathname = src_abs_pathname.relative_path_from(project_abs_pathname)
+
+  src_lines = File.readlines(src_path)
+  # lines that contain module declaration (e.g. "module ContentData")
+  module_lines = src_lines.select { |l| l =~ /module / }
```
Comment (bbfsdev) [inline Rakefile:88]: What is a module declaration? We don't have any modules, right?

Comment (bbfsdev) [inline Rakefile:88]: Maybe remove this part?

File: Rakefile [hunk 4]
```diff
@@ -25,11 +26,120 @@ task :uninstall do
...
+  spec_name = File.basename(src_path, '.rb') + '_spec.rb'
+  spec_path = File.join(spec_dir_name, spec_name)
+  if (File.exists?(spec_path))
+    puts "Spec #{spec_path} already exists. Do nothing. Exit.."
+    exit
+  end
+  
+  File.open(spec_path, 'w') do |f|
+    # NOTE Do not change indentation, cause it will break correct indentation in the spec file
+    f.write (
```
Comment (bbfsdev) [inline Rakefile:114]: What is being written to a file?

File: Rakefile [hunk 5]
```diff
@@ -25,11 +26,120 @@ task :uninstall do
...
+"# NOTE Code Coverage block must be issued before any of your application code is required
+if ENV['BBFS_COVERAGE']
+  require_relative '../spec_helper.rb'
+  SimpleCov.command_name '#{module_name}'
+end
+require 'rspec'
```
Comment (bbfsdev) [inline Rakefile:120]: Why do we need a require here? I don't understand how this task works.

File: Rakefile [hunk 6]
```diff
@@ -25,11 +26,120 @@ task :uninstall do
...
+    )
+    ends = Array.new
+    module_lines.each do |l|
+      ends << l.match(/^\s*/)[0] + "end\n"
+      f.write(l)
+    end
+    indent = if (ends.length > 1)
+               ends[1].match(/^\s*/)[0].length - ends[0].match(/^\s*/)[0].length
+             else
+               2
```
Comment (bbfsdev) [inline Rakefile:133]: 2? return 2?
confusing, add return please.

Comment (bbfsdev) [inline Rakefile:133]: What this 2 means?

**Comment by bbfsdev** [comment]
Review finished.

File: .simplecov (hunk 1 above)
Comment (genadyp) [inline .simplecov:12]: [lib/content_dat.rb](https://github.com/bbfsdev/bbfs/blob/master/lib/content_data.rb)

File: Rakefile (hunk 2 above)
Comment (genadyp) [inline Rakefile:66]: aggree.
add few lineto theworkfow document?

File: Rakefile (hunk 3 above)
Comment (genadyp) [inline Rakefile:88]: we have plenty of modules: 
e.g. [ContentData](https://github.com/bbfsdev/bbfs/blob/master/lib/content_data/content_data.rb#L5)

Comment (genadyp) [inline Rakefile:88]: I suggest to run:
`rake new_spec src=lib/email/email.rb`
To see how it works

File: Rakefile (hunk 4 above)
Comment (genadyp) [inline Rakefile:114]: Output of the this rake task is a template spec file.
I suggest to run:
`rake new_spec src=lib/email/email.rb`
To see how it works

File: Rakefile (hunk 5 above)
Comment (genadyp) [inline Rakefile:120]: Output of the this rake task is a template spec file.
I suggest to run:
`rake new_spec src=lib/email/email.rb`
To see how it works

File: Rakefile (hunk 6 above)
Comment (genadyp) [inline Rakefile:133]: 2 means an indent in the output spec file.
We do not need to `return` here cause in this case i think it causes unwanted return from the rake task.

**Pull request merged**